src/cli/keepassxc-cli.cpp

```diff
@@ -16,28 +16,157 @@
 #include <sanitizer/lsan_interface.h>
 #endif
 
+#if defined(USE_READLINE)
+#include <readline/history.h>
+#include <readline/readline.h>
+#endif
+
+class LineReader
+{
+public:
+    virtual ~LineReader() = default;
+    virtual QString readLine(QString prompt) = 0;
+    virtual bool isFinished() = 0;
+};
+
+class SimpleLineReader : public LineReader
+{
+public:
+    SimpleLineReader()
+        : inStream(stdin, QIODevice::ReadOnly),
+          outStream(stdout, QIODevice::WriteOnly),
+          finished(false)
+    {
+    }
+
+    QString readLine(QString prompt) override
+    {
+        outStream << prompt;
+        outStream.flush();
+        QString result = inStream.readLine();
+        if (result.isNull()) {
+            finished = true;
+        }
+        return result;
+    }
+
+    bool isFinished() override
+    {
+        return finished;
+    }
+
+private:
+    TextStream inStream;
+    TextStream outStream;
+    bool finished;
+};
+
+#if defined(USE_READLINE)
+class ReadlineLineReader : public LineReader
+{
+public:
+    ReadlineLineReader()
+        : finished(false)
+    {
+    }
+
+    QString readLine(QString prompt) override
+    {
+        char* result = readline(prompt.toStdString().c_str());
+        if (!result) {
+            finished = true;
+            return "";
+        }
+        add_history(result);
+        QString qstr(result);
+        free(result);
+        return qstr;
+    }
+
+    bool isFinished() override
+    {
+        return finished;
+    }
+
+private:
+    bool finished;
+};
+#endif
+
+void enterInteractiveMode(QStringList arguments) {
+
+    Open o;
+    arguments.removeFirst();
+    o.execute(arguments);
+
+    std::unique_ptr<LineReader> r;
+#if defined(USE_READLINE)
+    r.reset(new ReadlineLineReader);
+#else
+    r.reset(new SimpleLineReader);
+#endif
+
+    CommandParser parser;
+    QSharedPointer<Database> currentDatabase(o.currentDatabase);
+
+    QString command;
+    do {
+        TextStream errorTextStream(Utils::STDERR, QIODevice::WriteOnly);
+
+        QString prompt;
+        if (currentDatabase) {
+          prompt += QDir::current().relativeFilePath(currentDatabase->filePath());
+        }
+        prompt += "> ";
+        command = r->readLine(prompt);
+        if (r->isFinished()) {
+            return;
+        }
+
+        QStringList args = parser.parse(command);
+        if (args.empty()) {
+            continue;
+        }
+
+        Command* cmd = Command::getCommand(args[0]);
+        if (cmd == nullptr) {
+            errorTextStream << "Unknown command " << args[0] << "\n";
+            continue;
+        }
+
+        cmd->currentDatabase = currentDatabase;
+        cmd->execute(args);
+        currentDatabase = cmd->currentDatabase;
+    } while (true);
+}
+
 int main(int argc, char** argv)
 {
     if (!Crypto::init()) {
         qFatal("Fatal error while testing the cryptographic functions:\n%s", qPrintable(Crypto::errorString()));
-        return EXIT_FAILURE;
```

Comment:
@sjamesr I would leave the explicit exit just in case `qFatal` doesn't exit (looks like it can happen if the default message handlers are overridden)

Reply:
Reverted, sorry about that